### PR TITLE
FMI2: Parse type and unit definitions into `modelDescription`

### PR DESCRIPTION
**Pull request by manuelbb-upb**
Based on the changes in https://github.com/ThummeTo/FMICore.jl/pull/33, this pull request modifies parsing of a model definition such that the `modelDescription` of an `FMU2` actually holds non-trivial `typeDefinitons` and `unitDefinitions` if present.

There is also post-processing for the model variables to set the attributes according to a `declaredType` if applicable, i.e., if an attribute is not defined for the model variable but for the relevant SimpleType.

Lastly, there is a convenience function `fmi2GetUnit(model_description, model_variable)` to retrieve the unit corresponding to `model_variable.Real.unit` from `model_description.unitDefinitions`.

Marked as “draft” because:
* I am really new to this whole thing,
* the docs should be adapted,
* maybe these changes should be tested (automatically).

### CODE REVIEW

**Comment by ThummeTo** [comment]
Thank you very much, we will check the PRs and merge if everything is ok.
Much appreciated 👍 

@JoKircher

**Comment by JoKircher** [comment]
Sure, I will give it a look

File: src/FMI2/ext.jl [hunk 1]
```diff
@@ -1179,7 +1179,27 @@ function fmi2GetUnit(mv::fmi2ScalarVariable)
...
 """
+   fmi2GetUnit(md::fmi2ModelDescription, mv::fmi2ScalarVariable)
 
+Returns the `unit` of the corresponding model variable `mv` as a `fmi2Unit` if it is
+defined in `md.unitDefinitions`.
+
+# Source
+- FMISpec2.0.2 Link: [https://fmi-standard.org/](https://fmi-standard.org/)
+- FMISpec2.0.2: 2.2.7  Definition of Model Variables (ModelVariables)
+"""
```
Comment (JoKircher) [inline src/FMI2/ext.jl:1190]:
```suggestion
"""

   fmi2GetUnit(md::fmi2ModelDescription, mv::fmi2ScalarVariable)

Returns the `unit` of the corresponding model variable `mv` as a `fmi2Unit` if it is
defined in `md.unitDefinitions`.

# Arguments
- `md::fmi2ModelDescription`: Struct which provides the static information of ModelVariables.
- `mv::fmi2ScalarVariable`: The “ModelVariables” element consists of an ordered set of “ScalarVariable” elements. A “ScalarVariable” represents a variable of primitive type, like a real or integer variable.

# Source
- FMISpec2.0.2 Link: [https://fmi-standard.org/](https://fmi-standard.org/)
- FMISpec2.0.2: 2.2.7  Definition of Model Variables (ModelVariables)
"""
```

File: src/FMI2/md.jl [hunk 2]
```diff
@@ -171,6 +174,71 @@ function getDerivativeIndices(node::EzXML.Node)
...
+        if haskey(defnode, "start")
+            typenode.start = parseReal(defnode["start"])
+        end
+        if haskey(defnode, "derivative")
+            typenode.derivative = parse(UInt, defnode["derivative"])
+        end
+    elseif typename == "String"
+        typenode = fmi2ModelDescriptionString()
+        if haskey(defnode, "start")
+            targetArray[index].String.start = defnode["start"]
```
Comment (JoKircher) [inline src/FMI2/md.jl:216]:
```suggestion
            typenode.start = defnode["start"]
```

File: src/FMI2/md.jl [hunk 3]
```diff
@@ -171,6 +174,71 @@ function getDerivativeIndices(node::EzXML.Node)
...
+        end
+        # ToDo: remaining attributes
+    elseif typename == "Boolean"
+        typenode = fmi2ModelDescriptionBoolean()
+        if haskey(defnode, "start")
+            targetArray[index].Boolean.start = parseFMI2Boolean(defnode["start"])
```
Comment (JoKircher) [inline src/FMI2/md.jl:222]:
```suggestion
            typenode.start = parseFMI2Boolean(defnode["start"])
```

File: src/FMI2/md.jl [hunk 4]
```diff
@@ -171,6 +174,71 @@ function getDerivativeIndices(node::EzXML.Node)
...
+        end
+        # ToDo: remaining attributes
+    elseif typename == "Integer"
+        typenode = fmi2ModelDescriptionInteger()
+        if haskey(defnode, "start")
+            targetArray[index].Integer.start = parseInteger(defnode["start"])
```
Comment (JoKircher) [inline src/FMI2/md.jl:228]:
```suggestion
            typenode.start = parseInteger(defnode["start"])
```

File: src/FMI2/md.jl [hunk 5]
```diff
@@ -299,6 +315,68 @@ function parseModelVariables(nodes::EzXML.Node, md::fmi2ModelDescription)
     scalarVariables
 end
 
+setDefaultsWithSimpleType!(typenode, simpleType)=nothing
+
+function setDefaultsWithSimpleType!(typenode::fmi2ModelDescriptionReal, simpleType)
```
Comment (JoKircher) [inline src/FMI2/md.jl:320]: some documentation would be nice!

File: src/FMI2/md.jl [hunk 6]
```diff
@@ -299,6 +315,68 @@ function parseModelVariables(nodes::EzXML.Node, md::fmi2ModelDescription)
...
+            if !isnothing(src)
+                setfield!(typenode, attr, src)
+            end
+        end
+    end
+   
+    return nothing
+end
+
+function setDefaultsWithSimpleType!(typenode::fmi2ModelDescriptionInteger, simpleType)
```
Comment (JoKircher) [inline src/FMI2/md.jl:337]: Here too

File: src/FMI2/md.jl [hunk 7]
```diff
@@ -299,6 +315,68 @@ function parseModelVariables(nodes::EzXML.Node, md::fmi2ModelDescription)
...
+   
+    return nothing
+end
+
+# TODO
+# * decide if the fallback applies to the :start attribute, too (we currently ignore it)
+# * once all attributes are implemented for “Enumeration”, define `setDefaultsWithSimpleType`
+#   for `fmi2ModelDescriptionEnumeration`.
+
+function setDefaultsFromDeclaredType!(scalarVariables, simpleTypes)
```
Comment (JoKircher) [inline src/FMI2/md.jl:359]: and here

**Review by JoKircher — CHANGES_REQUESTED** [review]
there was a bug with targetArray, that failed the tests.

FYI, if you want to test your changes, you can add them to your environment with:
```
]
dev "path/to/your/package"
```

File: src/FMI2/md.jl (hunk 2 above)
Comment (manuelbb-upb) [reply src/FMI2/md.jl:216]: You are right, this is an historic artifact. Sorry!

**Comment by manuelbb-upb** [comment]
Thanks for the review and sorry for not running the unit tests. Will submit a commit soon adding some documentation.

**Pull request merged**